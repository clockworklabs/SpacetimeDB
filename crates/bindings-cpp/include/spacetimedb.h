#pragma once

/**
 * SpacetimeDB C++20 Module API - Single Include Header
 * 
 * Complete, optimized API for building SpacetimeDB modules with support for:
 * - Tables with field constraints
 * - BSATN serialization
 * - Reducers (user-defined and lifecycle)
 * - Database operations
 * - Logging and debugging
 * 
 * =============================================================================
 * USAGE EXAMPLES
 * =============================================================================
 * 
 * Table Definition:
 *    struct User {
 *        uint32_t id;
 *        std::string username;
 *        std::string email;
 *    };
 *    SPACETIMEDB_STRUCT(User, id, username, email)
 *    SPACETIMEDB_TABLE(User, "users", Public,
 *        PrimaryKeyAutoInc(id),
 *        Unique(email)
 *    )
 * 
 * Lifecycle Reducers:
 *    SPACETIMEDB_INIT(init_module) {
 *        LOG_INFO("Module initialized!");
 *    }
 * 
 * User Reducers:
 *    void add_user(ReducerContext ctx, std::string username, std::string email) {
 *        User user{0, username, email}; // id auto-generated
 *        ctx.db.table<User>("users").insert(user);
 *    }
 *    REGISTER_REDUCER(add_user, add_user)
 */

// =============================================================================
// STANDARD LIBRARY DEPENDENCIES
// =============================================================================

#include <cstdint>
#include <string>
#include <vector>
#include <optional>
#include <array>
#include <string_view>
#include <concepts>
#include <functional>
#include <type_traits>
#include <span>

// =============================================================================
// CORE SPACETIMEDB TYPE SYSTEM
// =============================================================================

// Core types and BSATN support
#include "spacetimedb/bsatn/types.h"
#include "spacetimedb/bsatn/timestamp.h"
#include "spacetimedb/bsatn/time_duration.h"
#include "spacetimedb/bsatn/writer.h"
#include "spacetimedb/bsatn/reader.h"
#include "spacetimedb/bsatn/algebraic_type.h"
#include "spacetimedb/bsatn/traits.h"
#include "spacetimedb/bsatn/sum_type.h"
#include "spacetimedb/bsatn/schedule_at.h"

// Autogenerated types for module definition
#include "spacetimedb/internal/autogen/SumType.g.h"  // Complete SumType definition first
#include "spacetimedb/internal/autogen/RawModuleDefV9.g.h"
#include "spacetimedb/internal/autogen/RawTableDefV9.g.h"
#include "spacetimedb/internal/autogen/Typespace.g.h"
#include "spacetimedb/internal/autogen/ProductType.g.h"
#include "spacetimedb/internal/autogen/ProductTypeElement.g.h"
#include "spacetimedb/internal/autogen/TableAccess.g.h"
#include "spacetimedb/internal/autogen/TableType.g.h"

// =============================================================================
// MODULE SYSTEM INFRASTRUCTURE
// =============================================================================

// Core module infrastructure
#include "spacetimedb/internal/Module.h"
#include "spacetimedb/internal/field_registration.h"

// Logging system
#include "spacetimedb/logger.h"

// =============================================================================
// TABLE AND CONSTRAINT SYSTEM
// =============================================================================

// Type extensions for special types
#include "spacetimedb/bsatn/type_extensions.h"

// Table operations and constraint support
#include "spacetimedb/table.h"
#include "spacetimedb/database.h"
#include "spacetimedb/table_with_constraints.h"

// =============================================================================
// REDUCER SYSTEM
// =============================================================================

// Reducer context and macros
#include "spacetimedb/reducer_context.h"
#include "spacetimedb/reducer_macros.h"
#include "spacetimedb/reducer_error.h"

// Client visibility filters
#include "spacetimedb/client_visibility_filter.h"
#include "spacetimedb/rls.h"

// =============================================================================
// CONVENIENCE ALIASES AND COMPATIBILITY
// =============================================================================

// Convenience aliases for table access levels
namespace SpacetimeDb {
    // Re-export table access constants for convenience
    constexpr auto Public = Internal::TableAccess::Public;
    constexpr auto Private = Internal::TableAccess::Private;
}

// =============================================================================
// UNIFIED MACRO SYSTEM
// =============================================================================

// Use unified macro system
#include "spacetimedb/macros.h"


// =============================================================================
// MODULE DEFINITION UTILITIES
// =============================================================================

// Forward declarations for module system compatibility
namespace spacetimedb {
    // Template specialization support
    template<typename T> struct table_metadata;
    
    namespace internal {
        /**
         * @brief Serialize the current module definition to BSATN
         * @return Serialized module definition
         */
        inline std::vector<uint8_t> serialize_module_def() {
            auto& module_def = SpacetimeDb::Internal::Module::GetModuleDef();
            return module_def.serialize();
        }
    }

    // =============================================================================
    // CONVENIENCE TYPE ALIASES
    // =============================================================================
    
    // Vector type aliases for SpacetimeDB types
    using VecTimeDuration = std::vector<::SpacetimeDb::TimeDuration>;
}

// =============================================================================
// DOCUMENTATION AND USAGE NOTES
// =============================================================================

/**
 * CONSTRAINT TYPES REFERENCE:
 * 
 * Basic Constraints:
 * - PrimaryKey(field) - Primary key constraint (unique, non-null)
 * - Unique(field) - Unique constraint (allows null)
 * - Index(field) - Index for fast lookups
 * - AutoInc(field) - Auto-incrementing field
 * 
 * Combination Constraints:
 * - PrimaryKeyAutoInc(field) - Primary key with auto-increment
 * - UniqueAutoInc(field) - Unique field with auto-increment  
 * - IndexAutoInc(field) - Indexed field with auto-increment
 * 
 * LIFECYCLE REDUCERS:
 * - SPACETIMEDB_INIT(name) - Called when module is initialized
 * - SPACETIMEDB_CLIENT_CONNECTED(name) - Called when client connects
 * - SPACETIMEDB_CLIENT_DISCONNECTED(name) - Called when client disconnects
 * 
 * LOGGING LEVELS:
 * - LOG_DEBUG(msg) - Debug information
 * - LOG_INFO(msg) - General information
 * - LOG_WARN(msg) - Warning messages
 * - LOG_ERROR(msg) - Error messages
 * - LOG_PANIC(msg) - Critical errors (may terminate)
 */

// =============================================================================
// BSATN IMPLEMENTATION INCLUDES
// =============================================================================

// Include BSATN implementation files after all headers are defined
#include "spacetimedb/bsatn/types_impl.h"
#include "spacetimedb/bsatn/schedule_at_impl.h"
#include "spacetimedb/enum_macro.h"