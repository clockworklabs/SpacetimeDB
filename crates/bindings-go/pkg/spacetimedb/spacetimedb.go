// Package spacetimedb provides Go bindings for SpacetimeDB.
//
// This package includes core types, reducer framework, schema management,
// BSATN serialization, and utilities for building SpacetimeDB modules in Go.
//
// Core Types:
//   - Identity: Player identity
//   - Timestamp: Point in time
//   - TimeDuration: Duration of time
//   - ScheduleAt: Scheduling specification
//
// Reducer Framework:
//   - ReducerFunction: Interface for reducers
//   - LifecycleFunction: Interface for lifecycle functions
//   - ReducerRegistry: Registry for managing reducers
//   - GenericReducer: Simple reducer implementation
//   - GenericLifecycleFunction: Simple lifecycle function implementation
//
// Schema Management:
//   - TableInfo: Table definition structure
//   - Column: Column definition structure
//   - Index: Index definition structure
//   - TableRegistry: Registry for table definitions
//
// BSATN Serialization:
//   - High-performance binary serialization format
//   - Support for all primitive types (u8, u16, u32, u64, i8, i16, i32, i64, f32, f64, bool, string, bytes)
//   - Advanced collections (arrays, maps, optionals)
//   - Generic encoding/decoding functions
//   - Size calculation utilities
//
// Example usage:
//
//	import "github.com/clockworklabs/SpacetimeDB/crates/bindings-go/pkg/spacetimedb"
//
//	// Use core types
//	identity := spacetimedb.NewIdentity([16]byte{...})
//	timestamp := spacetimedb.NewTimestampFromTime(time.Now())
//
//	// Create table schema
//	table := spacetimedb.NewTableInfo("users")
//	table.Columns = []spacetimedb.TableColumn{
//		spacetimedb.NewPrimaryKeyColumn("id", spacetimedb.TypeU32),
//		spacetimedb.NewColumn("name", spacetimedb.TypeString),
//	}
//	spacetimedb.GlobalRegisterTable(table)
//
//	// Create and register a reducer
//	myReducer := spacetimedb.NewGenericReducer("my_reducer", "Does something", func(ctx *spacetimedb.ReducerContext, args []byte) spacetimedb.ReducerResult {
//		// Your reducer logic here
//		return spacetimedb.NewSuccessResult()
//	})
//	spacetimedb.RegisterReducer(myReducer)
package spacetimedb

// Re-export all components for convenience
import (
	"github.com/clockworklabs/SpacetimeDB/crates/bindings-go/pkg/spacetimedb/bsatn"
	"github.com/clockworklabs/SpacetimeDB/crates/bindings-go/pkg/spacetimedb/reducers"
	"github.com/clockworklabs/SpacetimeDB/crates/bindings-go/pkg/spacetimedb/schema"
	"github.com/clockworklabs/SpacetimeDB/crates/bindings-go/pkg/spacetimedb/types"
)

// Core SpacetimeDB types
type (
	// Identity represents a unique player identity
	Identity = types.Identity

	// Timestamp represents a point in time
	Timestamp = types.Timestamp

	// TimeDuration represents a duration of time
	TimeDuration = types.TimeDuration

	// ScheduleAt represents when a scheduled reducer should be executed
	ScheduleAt = types.ScheduleAt
)

// Reducer framework types
type (
	// LifecycleType represents the type of lifecycle event
	LifecycleType = reducers.LifecycleType

	// ReducerContext contains information about the current reducer execution
	ReducerContext = reducers.ReducerContext

	// ReducerResult represents the result of a reducer execution
	ReducerResult = reducers.ReducerResult

	// ReducerFunction represents a function that can be called as a reducer
	ReducerFunction = reducers.ReducerFunction

	// LifecycleFunction represents a function that can be called for lifecycle events
	LifecycleFunction = reducers.LifecycleFunction

	// GenericReducer provides a simple implementation of ReducerFunction
	GenericReducer = reducers.GenericReducer

	// GenericLifecycleFunction provides a simple implementation of LifecycleFunction
	GenericLifecycleFunction = reducers.GenericLifecycleFunction

	// ReducerRegistry manages all registered reducers and lifecycle functions
	ReducerRegistry = reducers.ReducerRegistry

	// ReducerMetrics tracks performance metrics for reducers
	ReducerMetrics = reducers.ReducerMetrics
)

// Schema framework types
type (
	// TableInfo contains metadata about a SpacetimeDB table
	TableInfo = schema.TableInfo

	// TableColumn represents a table column definition
	TableColumn = schema.Column

	// TableIndex represents a table index definition
	TableIndex = schema.Index

	// TableRegistry manages a collection of table definitions
	TableRegistry = schema.TableRegistry

	// SchemaTableID represents a unique identifier for a SpacetimeDB table in schema
	SchemaTableID = schema.TableID

	// ColumnID represents a unique identifier for a table column
	ColumnID = schema.ColumnID

	// SchemaIndexID represents a unique identifier for a table index in schema
	SchemaIndexID = schema.IndexID

	// IndexType represents the type of index
	IndexType = schema.IndexType

	// TableRegistrationOptions configures table registration behavior
	TableRegistrationOptions = schema.RegistrationOptions

	// TableRegistryStats contains statistics about a table registry
	TableRegistryStats = schema.RegistryStats
)

// Lifecycle constants
const (
	LifecycleInit       = reducers.LifecycleInit
	LifecycleUpdate     = reducers.LifecycleUpdate
	LifecycleConnect    = reducers.LifecycleConnect
	LifecycleDisconnect = reducers.LifecycleDisconnect
)

// Schema constants - SpacetimeDB type names
const (
	TypeU8           = schema.TypeU8
	TypeU16          = schema.TypeU16
	TypeU32          = schema.TypeU32
	TypeU64          = schema.TypeU64
	TypeU128         = schema.TypeU128
	TypeU256         = schema.TypeU256
	TypeI8           = schema.TypeI8
	TypeI16          = schema.TypeI16
	TypeI32          = schema.TypeI32
	TypeI64          = schema.TypeI64
	TypeI128         = schema.TypeI128
	TypeI256         = schema.TypeI256
	TypeF32          = schema.TypeF32
	TypeF64          = schema.TypeF64
	TypeBool         = schema.TypeBool
	TypeString       = schema.TypeString
	TypeBytes        = schema.TypeBytes
	TypeIdentity     = schema.TypeIdentity
	TypeTimestamp    = schema.TypeTimestamp
	TypeTimeDuration = schema.TypeTimeDuration
	TypeScheduleAt   = schema.TypeScheduleAt
)

// Index type constants
const (
	IndexTypeBTree  = schema.IndexTypeBTree
	IndexTypeHash   = schema.IndexTypeHash
	IndexTypeDirect = schema.IndexTypeDirect
)

// Core type constructors
var (
	// Identity constructors
	NewIdentity = types.NewIdentity

	// Timestamp constructors
	NewTimestamp         = types.NewTimestamp
	NewTimestampFromTime = types.NewTimestampFromTime

	// TimeDuration constructors
	NewTimeDuration             = types.NewTimeDuration
	NewTimeDurationFromDuration = types.NewTimeDurationFromDuration

	// ScheduleAt constructors
	NewScheduleAtTime     = types.NewScheduleAtTime
	NewScheduleAtInterval = types.NewScheduleAtInterval
)

// Reducer result constructors
var (
	NewSuccessResult            = reducers.NewSuccessResult
	NewSuccessResultWithMessage = reducers.NewSuccessResultWithMessage
	NewErrorResult              = reducers.NewErrorResult
	NewErrorResultWithMessage   = reducers.NewErrorResultWithMessage
)

// Reducer and lifecycle constructors
var (
	NewGenericReducer           = reducers.NewGenericReducer
	NewGenericLifecycleFunction = reducers.NewGenericLifecycleFunction
	NewReducerRegistry          = reducers.NewReducerRegistry
	NewReducerMetrics           = reducers.NewReducerMetrics
)

// Schema constructors
var (
	// Table constructors
	NewTableInfo     = schema.NewTableInfo
	NewTableRegistry = schema.NewTableRegistry

	// Column constructors
	NewColumn           = schema.NewColumn
	NewPrimaryKeyColumn = schema.NewPrimaryKeyColumn
	NewAutoIncColumn    = schema.NewAutoIncColumn

	// Index constructors
	NewIndex       = schema.NewIndex
	NewBTreeIndex  = schema.NewBTreeIndex
	NewUniqueIndex = schema.NewUniqueIndex

	// Options constructor
	DefaultRegistrationOptions = schema.DefaultRegistrationOptions
)

// Global registry functions - Reducers
var (
	RegisterReducer           = reducers.RegisterReducer
	RegisterLifecycleFunction = reducers.RegisterLifecycleFunction
	GetReducer                = reducers.GetReducer
	GetLifecycleFunction      = reducers.GetLifecycleFunction
)

// Global registry functions - Tables
var (
	GlobalRegisterTable     = schema.GlobalRegister
	GlobalRegisterAllTables = schema.GlobalRegisterAll
	GlobalGetTable          = schema.GlobalGetTable
	GlobalGetTableByID      = schema.GlobalGetTableByID
	GlobalMustGetTable      = schema.GlobalMustGetTable
	GlobalHasTable          = schema.GlobalHasTable
	GlobalGetAllTables      = schema.GlobalGetAllTables
	GlobalGetTableNames     = schema.GlobalGetTableNames
	GlobalTableCount        = schema.GlobalCount
	GlobalClearTables       = schema.GlobalClear
	GlobalValidateAllTables = schema.GlobalValidateAll
	GlobalGetTableStats     = schema.GlobalGetStats
	GetGlobalTableRegistry  = schema.GetGlobalRegistry
)

// BSATN serialization functions
var (
	// Primitive type encoders
	BsatnEncodeU8     = bsatn.EncodeU8
	BsatnEncodeU16    = bsatn.EncodeU16
	BsatnEncodeU32    = bsatn.EncodeU32
	BsatnEncodeU64    = bsatn.EncodeU64
	BsatnEncodeI8     = bsatn.EncodeI8
	BsatnEncodeI16    = bsatn.EncodeI16
	BsatnEncodeI32    = bsatn.EncodeI32
	BsatnEncodeI64    = bsatn.EncodeI64
	BsatnEncodeF32    = bsatn.EncodeF32
	BsatnEncodeF64    = bsatn.EncodeF64
	BsatnEncodeBool   = bsatn.EncodeBool
	BsatnEncodeString = bsatn.EncodeString
	BsatnEncodeBytes  = bsatn.EncodeBytes

	// Primitive type decoders
	BsatnDecodeU8     = bsatn.DecodeU8
	BsatnDecodeU16    = bsatn.DecodeU16
	BsatnDecodeU32    = bsatn.DecodeU32
	BsatnDecodeU64    = bsatn.DecodeU64
	BsatnDecodeI8     = bsatn.DecodeI8
	BsatnDecodeI16    = bsatn.DecodeI16
	BsatnDecodeI32    = bsatn.DecodeI32
	BsatnDecodeI64    = bsatn.DecodeI64
	BsatnDecodeF32    = bsatn.DecodeF32
	BsatnDecodeF64    = bsatn.DecodeF64
	BsatnDecodeBool   = bsatn.DecodeBool
	BsatnDecodeString = bsatn.DecodeString
	BsatnDecodeBytes  = bsatn.DecodeBytes

	// Array encoders/decoders
	BsatnEncodeU32Array    = bsatn.EncodeU32Array
	BsatnDecodeU32Array    = bsatn.DecodeU32Array
	BsatnEncodeStringArray = bsatn.EncodeStringArray
	BsatnDecodeStringArray = bsatn.DecodeStringArray
	BsatnEncodeF64Array    = bsatn.EncodeF64Array
	BsatnDecodeF64Array    = bsatn.DecodeF64Array

	// Generic collection functions
	BsatnEncodeArray    = bsatn.EncodeArray[any]
	BsatnDecodeArray    = bsatn.DecodeArray[any]
	BsatnEncodeOptional = bsatn.EncodeOptional[any]
	BsatnDecodeOptional = bsatn.DecodeOptional[any]
	BsatnEncodeMap      = bsatn.EncodeMap[string, any]
	BsatnDecodeMap      = bsatn.DecodeMap[string, any]

	// Size calculation functions
	BsatnSizeU8          = bsatn.SizeU8
	BsatnSizeU16         = bsatn.SizeU16
	BsatnSizeU32         = bsatn.SizeU32
	BsatnSizeU64         = bsatn.SizeU64
	BsatnSizeI8          = bsatn.SizeI8
	BsatnSizeI16         = bsatn.SizeI16
	BsatnSizeI32         = bsatn.SizeI32
	BsatnSizeI64         = bsatn.SizeI64
	BsatnSizeF32         = bsatn.SizeF32
	BsatnSizeF64         = bsatn.SizeF64
	BsatnSizeBool        = bsatn.SizeBool
	BsatnSizeString      = bsatn.SizeString
	BsatnSizeBytes       = bsatn.SizeBytes
	BsatnSizeU32Array    = bsatn.SizeU32Array
	BsatnSizeStringArray = bsatn.SizeStringArray
	BsatnSizeF64Array    = bsatn.SizeF64Array

	// Utility functions
	BsatnToBytes              = bsatn.ToBytes
	BsatnFromBytes            = bsatn.FromBytes
	BsatnU32ArrayToBytes      = bsatn.U32ArrayToBytes
	BsatnU32ArrayFromBytes    = bsatn.U32ArrayFromBytes
	BsatnStringArrayToBytes   = bsatn.StringArrayToBytes
	BsatnStringArrayFromBytes = bsatn.StringArrayFromBytes
)

// Package information
const (
	Version = "0.3.0"
	Name    = "SpacetimeDB Go Bindings"
)

// GetVersion returns the current version of the bindings
func GetVersion() string {
	return Version
}

// GetPackageName returns the package name
func GetPackageName() string {
	return Name
}

// GetFeatures returns a list of supported features
func GetFeatures() []string {
	return []string{
		"Core Types (Identity, Timestamp, TimeDuration, ScheduleAt)",
		"Reducer Framework (Registration, Execution, Metrics)",
		"Schema Management (Tables, Columns, Indexes)",
		"BSATN Serialization (Binary, High-Performance)",
		"Collection Support (Arrays, Maps, Optionals)",
		"Global Registries (Thread-safe, Validation)",
		"JSON Serialization",
		"WASM Support",
	}
}
