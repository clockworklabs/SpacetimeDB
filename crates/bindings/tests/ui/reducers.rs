use spacetimedb::ReducerContext;

struct Test;

#[spacetimedb::reducer]
fn bad_type(_ctx: &ReducerContext, _a: Test) {}

#[spacetimedb::reducer]
fn bad_return_type(_ctx: &ReducerContext) -> Test {
    Test
}

#[spacetimedb::reducer]
fn lifetime<'a>(_ctx: &ReducerContext, _a: &'a str) {}

#[spacetimedb::reducer]
fn type_param<T>() {}

#[spacetimedb::reducer]
fn const_param<const X: u8>() {}

#[spacetimedb::reducer]
fn missing_ctx(_a: u8) {}

#[spacetimedb::reducer]
fn ctx_by_val(_ctx: ReducerContext, _a: u8) {}

#[spacetimedb::table(name = scheduled_table_missing_rows, scheduled(scheduled_table_missing_rows_reducer))]
struct ScheduledTableMissingRows {
    x: u8,
    y: u8,
}

// #[spacetimedb::reducer]
// fn scheduled_table_missing_rows_reducer(_ctx: &ReducerContext, _: &ScheduledTableMissingRows) {}

#[spacetimedb::table(name = scheduled_table, scheduled(scheduled_table_reducer))]
struct ScheduledTable {
    #[primary_key]
    #[auto_inc]
    scheduled_id: u64,
    scheduled_at: spacetimedb::ScheduleAt,
    x: u8,
    y: u8,
}

#[spacetimedb::reducer]
fn scheduled_table_reducer(_ctx: &ReducerContext, _x: u8, _y: u8) {}

fn main() {}
