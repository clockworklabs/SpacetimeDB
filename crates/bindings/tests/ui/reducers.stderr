error: type parameters are not allowed on reducers
  --> tests/ui/reducers.rs:17:15
   |
17 | fn type_param<T>() {}
   |               ^

error: const parameters are not allowed on reducers
  --> tests/ui/reducers.rs:20:16
   |
20 | fn const_param<const X: u8>() {}
   |                ^^^^^^^^^^^

error: scheduled table missing required columns; add these to your struct:
       #[primary_key]
       #[auto_inc]
       scheduled_id: u64,
       scheduled_at: spacetimedb::ScheduleAt,
  --> tests/ui/reducers.rs:28:59
   |
28 | #[spacetimedb::table(name = scheduled_table_missing_rows, scheduled(scheduled_table_missing_rows_reducer))]
   |                                                           ^^^^^^^^^

error[E0593]: function is expected to take 2 arguments, but it takes 3 arguments
  --> tests/ui/reducers.rs:37:56
   |
37 | #[spacetimedb::table(name = scheduled_table, scheduled(scheduled_table_reducer))]
   | -------------------------------------------------------^^^^^^^^^^^^^^^^^^^^^^^---
   | |                                                      |
   | |                                                      expected function that takes 2 arguments
   | required by a bound introduced by this call
...
48 | fn scheduled_table_reducer(_ctx: &ReducerContext, _x: u8, _y: u8) {}
   | ----------------------------------------------------------------- takes 3 arguments
   |
   = note: required for `for<'a> fn(&'a ReducerContext, u8, u8) {scheduled_table_reducer}` to implement `Reducer<'_, (ScheduledTable,)>`
   = note: required for `for<'a> fn(&'a ReducerContext, u8, u8) {scheduled_table_reducer}` to implement `ReducerForScheduledTable<'_, ScheduledTable>`
note: required by a bound in `scheduled_reducer_typecheck`
  --> src/rt.rs
   |
   | pub const fn scheduled_reducer_typecheck<'de, Row>(_x: impl ReducerForScheduledTable<'de, Row>)
   |                                                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `scheduled_reducer_typecheck`

error[E0277]: invalid reducer signature
 --> tests/ui/reducers.rs:6:4
  |
5 | #[spacetimedb::reducer]
  | ----------------------- required by a bound introduced by this call
6 | fn bad_type(_ctx: &ReducerContext, _a: Test) {}
  |    ^^^^^^^^ this reducer signature is not valid
  |
  = help: the trait `Reducer<'_, _>` is not implemented for fn item `for<'a> fn(&'a ReducerContext, Test) {bad_type}`
  = note:
  = note: reducer signatures must match the following pattern:
  = note:     `Fn(&ReducerContext, [T1, ...]) [-> Result<(), impl Display>]`
  = note: where each `Ti` type implements `SpacetimeType`.
  = note:
note: required by a bound in `register_reducer`
 --> src/rt.rs
  |
  | pub fn register_reducer<'a, A: Args<'a>, I: ReducerInfo>(_: impl Reducer<'a, A>) {
  |                                                                  ^^^^^^^^^^^^^^ required by this bound in `register_reducer`

error[E0277]: the reducer argument `Test` does not implement `SpacetimeType`
 --> tests/ui/reducers.rs:6:40
  |
6 | fn bad_type(_ctx: &ReducerContext, _a: Test) {}
  |                                        ^^^^ the trait `SpacetimeType` is not implemented for `Test`
  |
  = note: if you own the type, try adding `#[derive(SpacetimeType)]` to its definition
  = help: the following other types implement trait `SpacetimeType`:
            &T
            ()
            AlgebraicType
            AlgebraicTypeRef
            Arc<T>
            ArrayType
            Box<T>
            ColId
          and $N others
  = note: required for `Test` to implement `ReducerArg`

error[E0277]: invalid reducer signature
 --> tests/ui/reducers.rs:6:4
  |
5 | #[spacetimedb::reducer]
  | ----------------------- required by a bound introduced by this call
6 | fn bad_type(_ctx: &ReducerContext, _a: Test) {}
  |    ^^^^^^^^ this reducer signature is not valid
  |
  = help: the trait `Reducer<'_, _>` is not implemented for fn item `for<'a> fn(&'a ReducerContext, Test) {bad_type}`
  = note:
  = note: reducer signatures must match the following pattern:
  = note:     `Fn(&ReducerContext, [T1, ...]) [-> Result<(), impl Display>]`
  = note: where each `Ti` type implements `SpacetimeType`.
  = note:
note: required by a bound in `invoke_reducer`
 --> src/rt.rs
  |
  | pub fn invoke_reducer<'a, A: Args<'a>>(
  |        -------------- required by a bound in this function
  |     reducer: impl Reducer<'a, A>,
  |                   ^^^^^^^^^^^^^^ required by this bound in `invoke_reducer`

error[E0277]: invalid reducer signature
 --> tests/ui/reducers.rs:9:4
  |
8 | #[spacetimedb::reducer]
  | ----------------------- required by a bound introduced by this call
9 | fn bad_return_type(_ctx: &ReducerContext) -> Test {
  |    ^^^^^^^^^^^^^^^ this reducer signature is not valid
  |
  = help: the trait `Reducer<'_, _>` is not implemented for fn item `for<'a> fn(&'a ReducerContext) -> Test {bad_return_type}`
  = note:
  = note: reducer signatures must match the following pattern:
  = note:     `Fn(&ReducerContext, [T1, ...]) [-> Result<(), impl Display>]`
  = note: where each `Ti` type implements `SpacetimeType`.
  = note:
note: required by a bound in `register_reducer`
 --> src/rt.rs
  |
  | pub fn register_reducer<'a, A: Args<'a>, I: ReducerInfo>(_: impl Reducer<'a, A>) {
  |                                                                  ^^^^^^^^^^^^^^ required by this bound in `register_reducer`

error[E0277]: `Test` is not a valid reducer return type
 --> tests/ui/reducers.rs:9:46
  |
9 | fn bad_return_type(_ctx: &ReducerContext) -> Test {
  |                                              ^^^^ the trait `IntoReducerResult` is not implemented for `Test`
  |
  = note: reducers cannot return values -- you can only return `()` or `Result<(), impl Display>`
  = help: the following other types implement trait `IntoReducerResult`:
            ()
            Result<(), E>

error[E0277]: invalid reducer signature
 --> tests/ui/reducers.rs:9:4
  |
8 | #[spacetimedb::reducer]
  | ----------------------- required by a bound introduced by this call
9 | fn bad_return_type(_ctx: &ReducerContext) -> Test {
  |    ^^^^^^^^^^^^^^^ this reducer signature is not valid
  |
  = help: the trait `Reducer<'_, _>` is not implemented for fn item `for<'a> fn(&'a ReducerContext) -> Test {bad_return_type}`
  = note:
  = note: reducer signatures must match the following pattern:
  = note:     `Fn(&ReducerContext, [T1, ...]) [-> Result<(), impl Display>]`
  = note: where each `Ti` type implements `SpacetimeType`.
  = note:
note: required by a bound in `invoke_reducer`
 --> src/rt.rs
  |
  | pub fn invoke_reducer<'a, A: Args<'a>>(
  |        -------------- required by a bound in this function
  |     reducer: impl Reducer<'a, A>,
  |                   ^^^^^^^^^^^^^^ required by this bound in `invoke_reducer`

error[E0277]: invalid reducer signature
  --> tests/ui/reducers.rs:23:4
   |
22 | #[spacetimedb::reducer]
   | ----------------------- required by a bound introduced by this call
23 | fn missing_ctx(_a: u8) {}
   |    ^^^^^^^^^^^ this reducer signature is not valid
   |
   = help: the trait `Reducer<'_, _>` is not implemented for fn item `fn(u8) {missing_ctx}`
   = note:
   = note: reducer signatures must match the following pattern:
   = note:     `Fn(&ReducerContext, [T1, ...]) [-> Result<(), impl Display>]`
   = note: where each `Ti` type implements `SpacetimeType`.
   = note:
note: required by a bound in `register_reducer`
  --> src/rt.rs
   |
   | pub fn register_reducer<'a, A: Args<'a>, I: ReducerInfo>(_: impl Reducer<'a, A>) {
   |                                                                  ^^^^^^^^^^^^^^ required by this bound in `register_reducer`

error[E0277]: the first argument of a reducer must be `&ReducerContext`
  --> tests/ui/reducers.rs:23:20
   |
23 | fn missing_ctx(_a: u8) {}
   |                    ^^ first argument must be `&ReducerContext`
   |
   = help: the trait `ReducerContextArg` is not implemented for `u8`
   = help: the trait `ReducerContextArg` is implemented for `&ReducerContext`

error[E0277]: invalid reducer signature
  --> tests/ui/reducers.rs:23:4
   |
22 | #[spacetimedb::reducer]
   | ----------------------- required by a bound introduced by this call
23 | fn missing_ctx(_a: u8) {}
   |    ^^^^^^^^^^^ this reducer signature is not valid
   |
   = help: the trait `Reducer<'_, _>` is not implemented for fn item `fn(u8) {missing_ctx}`
   = note:
   = note: reducer signatures must match the following pattern:
   = note:     `Fn(&ReducerContext, [T1, ...]) [-> Result<(), impl Display>]`
   = note: where each `Ti` type implements `SpacetimeType`.
   = note:
note: required by a bound in `invoke_reducer`
  --> src/rt.rs
   |
   | pub fn invoke_reducer<'a, A: Args<'a>>(
   |        -------------- required by a bound in this function
   |     reducer: impl Reducer<'a, A>,
   |                   ^^^^^^^^^^^^^^ required by this bound in `invoke_reducer`

error[E0277]: invalid reducer signature
  --> tests/ui/reducers.rs:26:4
   |
25 | #[spacetimedb::reducer]
   | ----------------------- required by a bound introduced by this call
26 | fn ctx_by_val(_ctx: ReducerContext, _a: u8) {}
   |    ^^^^^^^^^^ this reducer signature is not valid
   |
   = help: the trait `Reducer<'_, _>` is not implemented for fn item `fn(ReducerContext, u8) {ctx_by_val}`
   = note:
   = note: reducer signatures must match the following pattern:
   = note:     `Fn(&ReducerContext, [T1, ...]) [-> Result<(), impl Display>]`
   = note: where each `Ti` type implements `SpacetimeType`.
   = note:
note: required by a bound in `register_reducer`
  --> src/rt.rs
   |
   | pub fn register_reducer<'a, A: Args<'a>, I: ReducerInfo>(_: impl Reducer<'a, A>) {
   |                                                                  ^^^^^^^^^^^^^^ required by this bound in `register_reducer`

error[E0277]: the first argument of a reducer must be `&ReducerContext`
  --> tests/ui/reducers.rs:26:21
   |
26 | fn ctx_by_val(_ctx: ReducerContext, _a: u8) {}
   |                     ^^^^^^^^^^^^^^ first argument must be `&ReducerContext`
   |
   = help: the trait `ReducerContextArg` is not implemented for `ReducerContext`
   = help: the trait `ReducerContextArg` is implemented for `&ReducerContext`

error[E0277]: invalid reducer signature
  --> tests/ui/reducers.rs:26:4
   |
25 | #[spacetimedb::reducer]
   | ----------------------- required by a bound introduced by this call
26 | fn ctx_by_val(_ctx: ReducerContext, _a: u8) {}
   |    ^^^^^^^^^^ this reducer signature is not valid
   |
   = help: the trait `Reducer<'_, _>` is not implemented for fn item `fn(ReducerContext, u8) {ctx_by_val}`
   = note:
   = note: reducer signatures must match the following pattern:
   = note:     `Fn(&ReducerContext, [T1, ...]) [-> Result<(), impl Display>]`
   = note: where each `Ti` type implements `SpacetimeType`.
   = note:
note: required by a bound in `invoke_reducer`
  --> src/rt.rs
   |
   | pub fn invoke_reducer<'a, A: Args<'a>>(
   |        -------------- required by a bound in this function
   |     reducer: impl Reducer<'a, A>,
   |                   ^^^^^^^^^^^^^^ required by this bound in `invoke_reducer`
